File: DS3_611_BFS_py/BFS.py
from collections import deque 
n, m  = map(int, input().split())
board = []
vis = [ [0]*m for _ in range(n) ]

for i in range(n):
    board.append(list(map(int, input().split())))

dx = [0,0,1,-1]
dy = [1,-1,0,0]
picNum=0
maxArea =0 
for i in range(n): 
    for j in range(m): 
            if vis[i][j]==0 and board[i][j]==1 : 
                picNum+=1
                q = deque() 
                q.append((i,j))
                vis[i][j]=1
                area = 0 
                while q : 
                    area +=1
                    now = q.pop() 
                    for k in range(4):
                        nx = now[0] + dx[k] 
                        ny = now[1] + dy[k]
                        if 0 <= nx < n and 0 <= ny < m and board[nx][ny] ==1 and vis[nx][ny] ==0 : 
                            q.append((nx,ny))
                            vis[nx][ny] = 1 
                maxArea = max(maxArea, area)
print(picNum, maxArea, sep='\n')

File: DS3_611_BFS_cpp/bfs.cpp
// 611<BFS>[그림]1926_c++
#include <bits/stdc++.h>
using namespace std;
int board[501][501];
int vis[501][501];
int dx[4] = {0,0,1,-1};
int dy[4] = {1,-1,0,0};
int n,m;
int main(void){
    ios::sync_with_stdio(0);
    cin.tie(0);
    cin >> n >> m ; 
    for(int i =0 ; i < n; i ++)
        for(int j =0 ; j < m; j ++) 
            cin >> board[i][j];  
    int maxPic = 0 ;
    int picNum = 0 ;

    for(int i =0 ; i < n; i ++){
        for(int j =0 ; j < m; j ++) {
            if(board[i][j]==0 || vis[i][j] ) continue;
            picNum++;
            queue < pair<int, int> > q ;
            vis[i][j]=1;
            q.push({i,j});
            int area = 0 ; 
            while(!q.empty()){
                area ++ ;
                pair <int, int> now = q.front();
                q.pop();
                for(int k =0; k < 4 ; k++){
                    int nx = now.first + dx[k];
                    int ny = now.second + dy[k];
                    if(nx < 0 || nx >= n || ny < 0 || ny >= m) continue;  
                    if (board[nx][ny] !=1 || vis[nx][ny])continue;
                    q.push({nx,ny});
                    vis[nx][ny] = 1 ; 
                }            
            }
            maxPic = max(maxPic, area);
        }
    }
    cout << picNum << "\n" <<  maxPic; 
} 